lockfile/src/lib.rs

```diff
@@ -136,8 +136,8 @@ impl Iterator for LockfileFormatIter {
             0 => LockfileFormat::Npm,
             1 => LockfileFormat::Yarn,
             2 => LockfileFormat::Gem,
-            3 => LockfileFormat::Poetry,
-            4 => LockfileFormat::Pip,
+            3 => LockfileFormat::Pip,
+            4 => LockfileFormat::Poetry,
```

Comment:
Switching the order here has an implication. Before, when `poetry` came first, a `pyproject.toml` file would be _fully_ parsed by Poetry whereas now `pip` will only resolve the top-level dependencies and not consider any of the (optional?) Poetry dependency groups.

Plus, doesn't this change also mean that `pyproject.toml` files will almost exclusively be parsed by `pip` and never `poetry`? This question assumes that an environment seeking to parse a Python manifest file is almost certainly going to have `pip` installed _and may also_ have `poetry`.

Here is an example of the difference, using the `phylum-ci` project:

<details><summary>Details</summary>
<p>

```
../phylum-ci  10  18 on  main [✘?] via 🐳 desktop-linux is 📦 v0.28.1 via 🐍 v3.11.3 (delme_pip_test) took 5m49s
❯ cat pyproject.toml
[build-system]
# NOTE: Changes to the build system values should be inspected closely!
# Python 3.11 support was added in `poetry-core` v1.3.2
requires = ["poetry-core>=1.3.2"]
build-backend = "poetry.core.masonry.api"

[tool.poetry]
name = "phylum"
version = "0.28.1"
description = "Utilities for integrating Phylum into CI pipelines (and beyond)"
license = "GPL-3.0-or-later"
authors = ["Phylum, Inc. <engineering@phylum.io>"]
homepage = "https://phylum.io/"
repository = "https://github.com/phylum-dev/phylum-ci"
documentation = "https://docs.phylum.io/"
readme = "README.md"
keywords = ["dependency", "security", "CI", "integration"]
# Classifiers can be found here: https://pypi.org/classifiers/
# TODO: Update the "Development Status" as the project/package matures
classifiers = [
    "Development Status :: 4 - Beta",
    "Intended Audience :: Developers",
    "License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)",
    "Natural Language :: English",
    "Environment :: Console",
    "Topic :: Security",
    "Topic :: Software Development",
    "Topic :: Software Development :: Quality Assurance",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3.8",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: 3.11",
]
packages = [
    { include = "phylum", from = "src" },
    { include = "tests", format = "sdist" },
]

[tool.poetry.urls]
"Issue Tracker" = "https://github.com/phylum-dev/phylum-ci/issues"
"CI" = "https://github.com/phylum-dev/phylum-ci/actions"

[tool.poetry.scripts]
phylum-init = "phylum.init.cli:main"
phylum-ci = "phylum.ci.cli:script_main"

[tool.poetry.dependencies]
python = ">=3.8,<3.12"
requests = "*"
cryptography = "*"
packaging = "*"
"ruamel.yaml" = "*"
pathspec = "*"
rich = "*"

[tool.poetry.group.test]
optional = true

[tool.poetry.group.test.dependencies]
pytest = "*"
tomli = "*"
tox = "*"
dulwich = "*"

[tool.poetry.group.ci]
optional = true

[tool.poetry.group.ci.dependencies]
tox-gh-actions = "*"
pytest-github-actions-annotate-failures = "*"
python-semantic-release = "*"
rich-codex = "*"

[tool.poetry.group.qa]
optional = true

[tool.poetry.group.qa.dependencies]
types-requests = "*"
pre-commit = "*"
tox = "*"

---TRIMMED---

../phylum-ci  10  18 on  main [✘?] via 🐳 desktop-linux is 📦 v0.28.1 via 🐍 v3.11.3 (delme_pip_test)
❯ phylum --version
phylum v5.2.0

../phylum-ci  10  18 on  main [✘?] via 🐳 desktop-linux is 📦 v0.28.1 via 🐍 v3.11.3 (delme_pip_test)
❯ phylum parse pyproject.toml
Generating lockfile for manifest "pyproject.toml" using Poetry…
[
  {
    "name": "attrs",
    "version": "23.1.0",
    "type": "pypi"
  },
  {
    "name": "bleach",
    "version": "6.0.0",
    "type": "pypi"
  },
  {
    "name": "cachetools",
    "version": "5.3.0",
    "type": "pypi"
  },
  {
    "name": "certifi",
    "version": "2023.5.7",
    "type": "pypi"
  },
  {
    "name": "cffi",
    "version": "1.15.1",
    "type": "pypi"
  },
  {
    "name": "cfgv",
    "version": "3.3.1",
    "type": "pypi"
  },
  {
    "name": "chardet",
    "version": "5.1.0",
    "type": "pypi"
  },
  {
    "name": "charset-normalizer",
    "version": "3.1.0",
    "type": "pypi"
  },
  {
    "name": "click",
    "version": "8.1.3",
    "type": "pypi"
  },
  {
    "name": "click-log",
    "version": "0.4.0",
    "type": "pypi"
  },
  {
    "name": "colorama",
    "version": "0.4.6",
    "type": "pypi"
  },
  {
    "name": "commonmark",
    "version": "0.9.1",
    "type": "pypi"
  },
  {
    "name": "cryptography",
    "version": "40.0.2",
    "type": "pypi"
  },
  {
    "name": "distlib",
    "version": "0.3.6",
    "type": "pypi"
  },
  {
    "name": "docutils",
    "version": "0.20.1",
    "type": "pypi"
  },
  {
    "name": "dotty-dict",
    "version": "1.3.1",
    "type": "pypi"
  },
  {
    "name": "dulwich",
    "version": "0.21.5",
    "type": "pypi"
  },
  {
    "name": "exceptiongroup",
    "version": "1.1.1",
    "type": "pypi"
  },
  {
    "name": "filelock",
    "version": "3.12.0",
    "type": "pypi"
  },
  {
    "name": "gitdb",
    "version": "4.0.10",
    "type": "pypi"
  },
  {
    "name": "gitpython",
    "version": "3.1.31",
    "type": "pypi"
  },
  {
    "name": "identify",
    "version": "2.5.24",
    "type": "pypi"
  },
  {
    "name": "idna",
    "version": "3.4",
    "type": "pypi"
  },
  {
    "name": "importlib-metadata",
    "version": "6.6.0",
    "type": "pypi"
  },
  {
    "name": "importlib-resources",
    "version": "5.12.0",
    "type": "pypi"
  },
  {
    "name": "iniconfig",
    "version": "2.0.0",
    "type": "pypi"
  },
  {
    "name": "invoke",
    "version": "1.7.3",
    "type": "pypi"
  },
  {
    "name": "jaraco-classes",
    "version": "3.2.3",
    "type": "pypi"
  },
  {
    "name": "jeepney",
    "version": "0.8.0",
    "type": "pypi"
  },
  {
    "name": "jsonschema",
    "version": "4.17.3",
    "type": "pypi"
  },
  {
    "name": "keyring",
    "version": "23.13.1",
    "type": "pypi"
  },
  {
    "name": "levenshtein",
    "version": "0.21.0",
    "type": "pypi"
  },
  {
    "name": "more-itertools",
    "version": "9.1.0",
    "type": "pypi"
  },
  {
    "name": "nodeenv",
    "version": "1.8.0",
    "type": "pypi"
  },
  {
    "name": "packaging",
    "version": "23.1",
    "type": "pypi"
  },
  {
    "name": "pathspec",
    "version": "0.11.1",
    "type": "pypi"
  },
  {
    "name": "pkginfo",
    "version": "1.9.6",
    "type": "pypi"
  },
  {
    "name": "pkgutil-resolve-name",
    "version": "1.3.10",
    "type": "pypi"
  },
  {
    "name": "platformdirs",
    "version": "3.5.1",
    "type": "pypi"
  },
  {
    "name": "pluggy",
    "version": "1.0.0",
    "type": "pypi"
  },
  {
    "name": "pre-commit",
    "version": "3.3.2",
    "type": "pypi"
  },
  {
    "name": "pycparser",
    "version": "2.21",
    "type": "pypi"
  },
  {
    "name": "pygments",
    "version": "2.15.1",
    "type": "pypi"
  },
  {
    "name": "pyproject-api",
    "version": "1.5.1",
    "type": "pypi"
  },
  {
    "name": "pyrsistent",
    "version": "0.19.3",
    "type": "pypi"
  },
  {
    "name": "pytest",
    "version": "7.3.1",
    "type": "pypi"
  },
  {
    "name": "pytest-github-actions-annotate-failures",
    "version": "0.2.0",
    "type": "pypi"
  },
  {
    "name": "python-gitlab",
    "version": "3.14.0",
    "type": "pypi"
  },
  {
    "name": "python-semantic-release",
    "version": "7.33.5",
    "type": "pypi"
  },
  {
    "name": "pywin32-ctypes",
    "version": "0.2.0",
    "type": "pypi"
  },
  {
    "name": "pyyaml",
    "version": "6.0",
    "type": "pypi"
  },
  {
    "name": "rapidfuzz",
    "version": "3.0.0",
    "type": "pypi"
  },
  {
    "name": "readme-renderer",
    "version": "37.3",
    "type": "pypi"
  },
  {
    "name": "requests",
    "version": "2.31.0",
    "type": "pypi"
  },
  {
    "name": "requests-toolbelt",
    "version": "1.0.0",
    "type": "pypi"
  },
  {
    "name": "rfc3986",
    "version": "2.0.0",
    "type": "pypi"
  },
  {
    "name": "rich",
    "version": "12.6.0",
    "type": "pypi"
  },
  {
    "name": "rich-cli",
    "version": "1.8.0",
    "type": "pypi"
  },
  {
    "name": "rich-click",
    "version": "1.6.1",
    "type": "pypi"
  },
  {
    "name": "rich-codex",
    "version": "1.2.6",
    "type": "pypi"
  },
  {
    "name": "rich-rst",
    "version": "1.1.7",
    "type": "pypi"
  },
  {
    "name": "ruamel-yaml",
    "version": "0.17.26",
    "type": "pypi"
  },
  {
    "name": "ruamel-yaml-clib",
    "version": "0.2.7",
    "type": "pypi"
  },
  {
    "name": "secretstorage",
    "version": "3.3.3",
    "type": "pypi"
  },
  {
    "name": "semver",
    "version": "2.13.0",
    "type": "pypi"
  },
  {
    "name": "setuptools",
    "version": "67.8.0",
    "type": "pypi"
  },
  {
    "name": "six",
    "version": "1.16.0",
    "type": "pypi"
  },
  {
    "name": "smmap",
    "version": "5.0.0",
    "type": "pypi"
  },
  {
    "name": "textual",
    "version": "0.1.18",
    "type": "pypi"
  },
  {
    "name": "tomli",
    "version": "2.0.1",
    "type": "pypi"
  },
  {
    "name": "tomlkit",
    "version": "0.11.8",
    "type": "pypi"
  },
  {
    "name": "tox",
    "version": "4.5.1",
    "type": "pypi"
  },
  {
    "name": "tox-gh-actions",
    "version": "3.1.1",
    "type": "pypi"
  },
  {
    "name": "tqdm",
    "version": "4.65.0",
    "type": "pypi"
  },
  {
    "name": "twine",
    "version": "3.8.0",
    "type": "pypi"
  },
  {
    "name": "types-requests",
    "version": "2.31.0.0",
    "type": "pypi"
  },
  {
    "name": "types-urllib3",
    "version": "1.26.25.13",
    "type": "pypi"
  },
  {
    "name": "typing-extensions",
    "version": "4.6.1",
    "type": "pypi"
  },
  {
    "name": "urllib3",
    "version": "2.0.2",
    "type": "pypi"
  },
  {
    "name": "virtualenv",
    "version": "20.23.0",
    "type": "pypi"
  },
  {
    "name": "webencodings",
    "version": "0.5.1",
    "type": "pypi"
  },
  {
    "name": "wheel",
    "version": "0.40.0",
    "type": "pypi"
  },
  {
    "name": "zipp",
    "version": "3.15.0",
    "type": "pypi"
  }
]

../phylum-ci  10  18 on  main [✘?] via 🐳 desktop-linux is 📦 v0.28.1 via 🐍 v3.11.3 (delme_pip_test) took 5s
❯ ~/dev/phylum/localdev/cli/target/debug/phylum --version
phylum v5.2.0-18-gac8983c

../phylum-ci  10  18 on  main [✘?] via 🐳 desktop-linux is 📦 v0.28.1 via 🐍 v3.11.3 (delme_pip_test)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse pyproject.toml
Generating lockfile for manifest "pyproject.toml" using Pip…
[
  {
    "name": "cryptography",
    "version": "40.0.2",
    "type": "pypi"
  },
  {
    "name": "packaging",
    "version": "23.1",
    "type": "pypi"
  },
  {
    "name": "pathspec",
    "version": "0.11.1",
    "type": "pypi"
  },
  {
    "name": "requests",
    "version": "2.31.0",
    "type": "pypi"
  },
  {
    "name": "rich",
    "version": "13.3.5",
    "type": "pypi"
  },
  {
    "name": "ruamel.yaml",
    "version": "0.17.26",
    "type": "pypi"
  },
  {
    "name": "certifi",
    "version": "2023.5.7",
    "type": "pypi"
  },
  {
    "name": "cffi",
    "version": "1.15.1",
    "type": "pypi"
  },
  {
    "name": "charset-normalizer",
    "version": "3.1.0",
    "type": "pypi"
  },
  {
    "name": "idna",
    "version": "3.4",
    "type": "pypi"
  },
  {
    "name": "markdown-it-py",
    "version": "2.2.0",
    "type": "pypi"
  },
  {
    "name": "Pygments",
    "version": "2.15.1",
    "type": "pypi"
  },
  {
    "name": "ruamel.yaml.clib",
    "version": "0.2.7",
    "type": "pypi"
  },
  {
    "name": "urllib3",
    "version": "2.0.2",
    "type": "pypi"
  },
  {
    "name": "mdurl",
    "version": "0.1.2",
    "type": "pypi"
  },
  {
    "name": "pycparser",
    "version": "2.21",
    "type": "pypi"
  }
]
```

</p>
</details>

The difference is that these top-level dependencies were parsed:

```
[tool.poetry.dependencies]
python = ">=3.8,<3.12"
requests = "*"
cryptography = "*"
packaging = "*"
"ruamel.yaml" = "*"
pathspec = "*"
rich = "*"
```

while these optional dependency groups were not:

```
[tool.poetry.group.test]
optional = true

[tool.poetry.group.test.dependencies]
pytest = "*"
tomli = "*"
tox = "*"
dulwich = "*"

[tool.poetry.group.ci]
optional = true

[tool.poetry.group.ci.dependencies]
tox-gh-actions = "*"
pytest-github-actions-annotate-failures = "*"
python-semantic-release = "*"
rich-codex = "*"

[tool.poetry.group.qa]
optional = true

[tool.poetry.group.qa.dependencies]
types-requests = "*"
pre-commit = "*"
tox = "*"
```

Reply:
Yes your summary is correct and this change in behavior is intentional.

If it was kept the other way around, you'd get failure when a non-poetry pyproject.toml is parsed with poetry so it wouldn't work at all.

Reply:
Further testing has revealed that parsing will switch from the `pyproject.toml` manifest to the `poetry.lock` lockfile, when that lockfile is in the same directory (as confirmed by [the lockfile generation docs](https://github.com/phylum-dev/cli/blob/main/docs/knowledge_base/lockfile-generation.md)). This is a good assumption for Poetry projects since they will almost exclusively have that file in the same directory.

The only scenario that could be envisioned where that is not the case is one where `poetry.lock` is not kept in version control. This could be the case for library developers, but the [Poetry docs](https://python-poetry.org/docs/basic-usage/#committing-your-poetrylock-file-to-version-control) were updated recently to suggest even library developers would find benefit in committing the lockfile to version control.

I can retract my "change requested" here...but do think this/these edge cases are worth enumerating.

Reply:
> Further testing has revealed that parsing will switch from the pyproject.toml manifest to the poetry.lock lockfile, when that lockfile is in the same directory (as confirmed by [the lockfile generation docs](https://github.com/phylum-dev/cli/blob/main/docs/knowledge_base/lockfile-generation.md)). This is a good assumption for Poetry projects since they will almost exclusively have that file in the same directory.

Yes, I'm sorry for not sharing the detail but that's the conclusion me and @kylewillmon came to too. For poetry the lockfile should be available.